File: android/app/src/main/java/app/farmsy/android/features/home/HomeScreen.kt
package app.farmsy.android.features.home

import android.Manifest
import android.content.Intent
import android.net.Uri
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.ChevronRight
import androidx.compose.material.icons.filled.LocationOff
import androidx.compose.material.icons.filled.Schedule
import androidx.compose.material.icons.outlined.FavoriteBorder
import androidx.compose.material.icons.outlined.LocationOn
import androidx.compose.material.icons.outlined.Notifications
import androidx.compose.material.icons.outlined.Person
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.farmsy.android.LocalFarms
import app.farmsy.android.LocalFavorites
import app.farmsy.android.LocalLocationHelper
import app.farmsy.android.LocalRequestAuth
import app.farmsy.android.LocalSession
import app.farmsy.android.R
import app.farmsy.android.core.FarmFilters
import app.farmsy.android.core.FarmPin
import app.farmsy.android.core.FarmsStore
import app.farmsy.android.core.ProductNearby
import app.farmsy.android.core.RecentReports
import app.farmsy.android.core.ReportStatus
import app.farmsy.android.core.SearchRadius
import app.farmsy.android.core.Seasons
import app.farmsy.android.core.ShoppingItems
import app.farmsy.android.features.main.AppTab
import app.farmsy.android.features.detail.minutesAgoLabel
import app.farmsy.android.features.main.LocalShell
import app.farmsy.android.features.whatsnew.SkeletonBox
import app.farmsy.android.ui.theme.AtmosphereBand
import app.farmsy.android.ui.ProductImage
import app.farmsy.android.ui.theme.Badge
import app.farmsy.android.ui.theme.CardShape
import app.farmsy.android.ui.theme.FarmsyColors
import app.farmsy.android.ui.theme.IconButton
import app.farmsy.android.ui.theme.ListRow
import app.farmsy.android.ui.theme.PillButton
import app.farmsy.android.ui.theme.PillSize
import app.farmsy.android.ui.theme.PillVariant
import app.farmsy.android.ui.theme.PlusLockCard
import app.farmsy.android.ui.theme.Radius
import app.farmsy.android.ui.theme.SearchField
import app.farmsy.android.ui.theme.SectionHeader
import app.farmsy.android.ui.theme.Space
import app.farmsy.android.ui.theme.TabBarInset
import app.farmsy.android.ui.theme.TextRole
import app.farmsy.android.ui.theme.TileShape
import app.farmsy.android.ui.theme.Wordmark
import app.farmsy.android.ui.theme.card
import app.farmsy.android.ui.theme.rememberTapHaptic
import app.farmsy.android.ui.theme.role
import app.farmsy.android.ui.theme.tapCard
import app.farmsy.android.ui.theme.ui
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.time.Duration
import java.time.Instant
import java.util.Calendar

/// "Where am I" for a screen: asks for the permission when it is missing, takes
/// a fix when it is granted. Mirrors iOS `locationManager.request()`.
@Composable
fun rememberLocationRequest(): () -> Unit {
    val locationHelper = LocalLocationHelper.current
    val launcher = rememberLauncherForActivityResult(ActivityResultContracts.RequestPermission()) { granted ->
        if (granted) locationHelper.request()
    }
    return {
        if (locationHelper.hasPermission()) locationHelper.request()
        else launcher.launch(Manifest.permission.ACCESS_FINE_LOCATION)
    }
}

/// Home — "what should I know today?" Personal, not a directory: what is
/// available near you, the farms you follow, and what Plus would tell you.
/// Profile lives behind the person button in the header.
@Composable
fun HomeScreen() {
    val context = LocalContext.current
    val session = LocalSession.current
    val farms = LocalFarms.current
    val favorites = LocalFavorites.current
    val locationHelper = LocalLocationHelper.current
    val shell = LocalShell.current
    val requestAuth = LocalRequestAuth.current
    val scope = rememberCoroutineScope()
    val tap = rememberTapHaptic()
    val requestLocation = rememberLocationRequest()

    val currentSession by session.session.collectAsState()
    val profile by session.profile.collectAsState()
    val isAuthenticated = currentSession != null
    val pins by farms.pins.collectAsState()
    val flagsLoaded by farms.flagsLoaded.collectAsState()
    val savedIds by favorites.osmIds.collectAsState()
    val location by locationHelper.location.collectAsState()
    val radiusKm by SearchRadius.km.collectAsState()
    val items by ShoppingItems.items.collectAsState()
    val language = remember { ShoppingItems.language(context) }

    var query by remember { mutableStateOf("") }
    var nearby by remember { mutableStateOf<List<ProductNearby>>(emptyList()) }
    var nearbyReady by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) { ShoppingItems.loadIfNeeded() }
    LaunchedEffect(Unit) { farms.loadFlagsIfNeeded() }
    LaunchedEffect(Unit) { Seasons.loadIfNeeded() }
    // Plus: the recent-reports feed, for "confirmed 18 min ago" on a tile.
    val plus = session.hasFullAccess
    LaunchedEffect(plus) { if (plus) RecentReports.refresh() }
    val recentReports by RecentReports.reports.collectAsState()
    val nearReports = remember(recentReports, plus, location, radiusKm, pins) {
        if (!plus || location == null) emptyList()
        else RecentReports.near(location, radiusKm, pins.associateBy { it.osmId })
    }
    val seasonItems by Seasons.items.collectAsState()
    val seasonMonth by Seasons.month.collectAsState()
    val seasonPicks = remember(seasonItems, seasonMonth) { Seasons.thisMonth(seasonItems, seasonMonth).take(6) }
    LaunchedEffect(Unit) { if (location == null && locationHelper.hasPermission()) locationHelper.request() }
    LaunchedEffect(location, radiusKm, items.size, flagsLoaded, pins.size) {
        val loc = location ?: return@LaunchedEffect
        if (!flagsLoaded || items.isEmpty() || pins.isEmpty()) return@LaunchedEffect
        val produce = farms.produceByOsm
        nearby = withContext(Dispatchers.Default) {
            FarmsStore.productsNearby(items, pins, produce, loc.latitude, loc.longitude, radiusKm)
        }
        nearbyReady = true
    }

    val greeting = run {
        val hour = Calendar.getInstance().get(Calendar.HOUR_OF_DAY)
        val base = stringResource(
            when (hour) { in 5..11 -> R.string.good_morning; in 12..17 -> R.string.good_afternoon; else -> R.string.good_evening }
        )
        val name = profile?.firstName?.trim().orEmpty()
        if (name.isEmpty()) base else "$base, $name"
    }
    val savedPins = remember(pins, savedIds, location) {
        farms.sortedByDistance(pins.filter { it.osmId in savedIds }, location)
    }

    Column(
        Modifier.fillMaxSize().background(FarmsyColors.cream).verticalScroll(rememberScrollState()),
    ) {
        // MARK: Header band
        Box(Modifier.fillMaxWidth().height(176.dp)) {
            AtmosphereBand(Modifier.fillMaxSize())
            Row(
                Modifier.fillMaxWidth().statusBarsPadding().padding(horizontal = Space.s4, vertical = Space.s2),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                IconButton(Icons.Outlined.Person, stringResource(R.string.profile), overMedia = true) { shell.openProfile() }
                Spacer(Modifier.weight(1f))
                Wordmark(onAtmosphere = true)
                Spacer(Modifier.weight(1f))
                var menu by remember { mutableStateOf(false) }
                Box {
                    IconButton(Icons.Outlined.LocationOn, stringResource(R.string.search_radius), overMedia = true) { menu = true }
                    DropdownMenu(expanded = menu, onDismissRequest = { menu = false }, containerColor = FarmsyColors.surface) {
                        SearchRadius.choices.forEach { km ->
                            DropdownMenuItem(
                                text = { Text("${km.toInt()} km", style = ui(16.sp), color = FarmsyColors.ink) },
                                leadingIcon = { if (km == radiusKm) Icon(Icons.Filled.Check, null, tint = FarmsyColors.ink, modifier = Modifier.size(18.dp)) },
                                onClick = { SearchRadius.set(context, km); menu = false },
                            )
                        }
                    }
                }
            }
        }

        Column(Modifier.padding(horizontal = Space.s4).padding(bottom = TabBarInset.content)) {
            // MARK: Greeting
            Column(
                Modifier.offset(y = -Space.s2).fillMaxWidth().card(),
                verticalArrangement = Arrangement.spacedBy(Space.s4),
            ) {
                Column(verticalArrangement = Arrangement.spacedBy(Space.s1)) {
                    Text(greeting, style = role(TextRole.HEADING), color = FarmsyColors.ink)
                    Text(
                        if (location == null) stringResource(R.string.home_local_food_near)
                        else stringResource(R.string.home_local_food_within_arg, radiusKm.toInt()),
                        style = role(TextRole.BODY_SM), color = FarmsyColors.inkMuted,
                    )
                }
                SearchField(query, { query = it }, stringResource(R.string.search_farms_or_products)) {
                    val q = query.trim()
                    if (q.isNotEmpty()) {
                        farms.searchText.value = q
                        shell.showTab(AppTab.MAP)
                    }
                }
                PillButton(stringResource(R.string.open_now_near_me), PillVariant.PRIMARY, PillSize.MEDIUM, block = true, icon = Icons.Filled.Schedule) {
                    farms.clearAllFilters()
                    farms.filterOpenNow.value = true
                    shell.showTab(AppTab.MAP)
                }
            }

            // MARK: Available near you
            SectionHeader(stringResource(R.string.available_near_you), stringResource(R.string.shopping) to { shell.showTab(AppTab.SHOPPING) })
            when {
                location == null -> InfoCard(
                    Icons.Filled.LocationOff, stringResource(R.string.where_are_you), stringResource(R.string.allow_location_to_see),
                    stringResource(R.string.allow) to requestLocation,
                )
                !nearbyReady -> Row(horizontalArrangement = Arrangement.spacedBy(Space.s3)) {
                    repeat(3) { SkeletonBox(cornerRadius = Radius.tile, modifier = Modifier.size(172.dp, 120.dp)) }
                }
                nearby.isEmpty() -> Text(
                    stringResource(R.string.home_no_farm_lists_arg, radiusKm.toInt()),
                    style = role(TextRole.BODY_SM), color = FarmsyColors.inkMuted, modifier = Modifier.fillMaxWidth().card(),
                )
                else -> Row(
                    Modifier.horizontalScroll(rememberScrollState()),
                    horizontalArrangement = Arrangement.spacedBy(Space.s3),
                ) {
                    nearby.take(10).forEach { p ->
                        Column(
                            Modifier.width(172.dp).background(FarmsyColors.surface, TileShape).tapCard {
                                scope.launch {
                                    farms.showProduct(p.item.label(language), p.item.terms, location, radiusKm)
                                    shell.showTab(AppTab.MAP)
                                }
                            }.padding(Space.s4),
                            verticalArrangement = Arrangement.spacedBy(Space.s2),
                        ) {
                            ProductImage(p.item.imageSlug, p.item.emoji, 64.dp)
                            Text(p.item.label(language), style = role(TextRole.SUBHEADING), color = FarmsyColors.ink, maxLines = 1, overflow = TextOverflow.Ellipsis)
                            Text(
                                stringResource(R.string.home_farms_km_arg, p.count, String.format("%.1f", p.nearestKm)),
                                style = role(TextRole.CAPTION), color = FarmsyColors.inkMuted, maxLines = 1,
                            )
                            // Plus: the newest "open" report within the radius that names
                            // this product. Free tiles say nothing about timing.
                            val hit = nearReports
                                .filter { it.status == ReportStatus.OPEN.wire && p.item.id in it.products }
                                .mapNotNull { it.report.instant }.maxOrNull()
                            if (hit != null) {
                                val mins = maxOf(0L, Duration.between(hit, Instant.now()).toMinutes()).toInt()
                                Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                                    Box(Modifier.size(6.dp).background(FarmsyColors.vividPositive, CircleShape))
                                    Text(
                                        stringResource(R.string.confirmed_arg, minutesAgoLabel(mins)),
                                        style = role(TextRole.CAPTION), color = FarmsyColors.positive, maxLines = 1,
                                    )
                                }
                            }
                        }
                    }
                }
            }

            // MARK: This week — season news. Renders nothing until the calendar has
            // loaded, so a missing endpoint costs a section, not a screen.
            if (seasonPicks.isNotEmpty()) {
                SectionHeader(stringResource(R.string.in_season_now), stringResource(R.string.see_all) to { shell.showTab(AppTab.DISCOVER) })
                Row(Modifier.horizontalScroll(rememberScrollState()), horizontalArrangement = Arrangement.spacedBy(Space.s3)) {
                    seasonPicks.forEach { item ->
                        Column(
                            Modifier.width(148.dp).background(FarmsyColors.surface, CardShape).tapCard {
                                scope.launch {
                                    farms.showProduct(item.label(language), item.terms, location, radiusKm)
                                    shell.showTab(AppTab.MAP)
                                }
                            }.padding(Space.s3),
                            verticalArrangement = Arrangement.spacedBy(Space.s2),
                        ) {
                            Box {
                                ProductImage(item.imageSlug, item.emoji, 124.dp, corner = Radius.tile)
                                if (item.isPeak(seasonMonth)) {
                                    Badge(stringResource(R.string.peak), fill = FarmsyColors.vivid, ink = FarmsyColors.ink, modifier = Modifier.align(Alignment.TopEnd).padding(Space.s2))
                                }
                            }
                            Text(item.label(language), style = role(TextRole.SUBHEADING), color = FarmsyColors.ink, maxLines = 1, overflow = TextOverflow.Ellipsis)
                            Text(stringResource(R.string.find_it_nearby), style = role(TextRole.CAPTION), color = FarmsyColors.inkMuted)
                        }
                    }
                }
            }

            // MARK: Your farms
            SectionHeader(
                stringResource(R.string.your_farms),
                if (isAuthenticated && savedPins.size > 3) stringResource(R.string.see_all) to { shell.showTab(AppTab.MAP) } else null,
            )
            when {
                !isAuthenticated -> InfoCard(
                    Icons.Outlined.FavoriteBorder, stringResource(R.string.keep_your_favourites), stringResource(R.string.sign_in_to_follow_farms),
                    stringResource(R.string.sign_in) to requestAuth,
                )
                savedPins.isEmpty() -> InfoCard(
                    Icons.Outlined.FavoriteBorder, stringResource(R.string.no_farms_followed_yet), stringResource(R.string.tap_heart_opening), null,
                )
                else -> Column(verticalArrangement = Arrangement.spacedBy(Space.s2)) {
                    savedPins.take(3).forEach { pin -> FarmRow(pin, location?.let { pin.distanceMeters(it.latitude, it.longitude) / 1000 }) { tap(); shell.openFarm(pin) } }
                }
            }

            // MARK: For you
            SectionHeader(stringResource(R.string.for_you))
            if (session.hasFullAccess) {
                Box(Modifier.fillMaxWidth().background(FarmsyColors.surface, CardShape)) {
                    ListRow(Icons.Outlined.Notifications, stringResource(R.string.product_alerts), subtitle = stringResource(R.string.product_alerts_sub)) {
                        context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("https://www.farmsy.app/alerts")))
                    }
                }
            } else {
                PlusLockCard(stringResource(R.string.plus_alerts_title), stringResource(R.string.plus_alerts_text), onUnlock = shell.openPlus)
            }
        }
    }
}

/// Icon, title + caption, optional small primary action — the "do this first" card.
@Composable
private fun InfoCard(icon: ImageVector, title: String, text: String, action: Pair<String, () -> Unit>?) {
    Row(Modifier.fillMaxWidth().card(), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(Space.s3)) {
        Icon(icon, null, tint = FarmsyColors.farmGreen, modifier = Modifier.size(20.dp))
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            Text(title, style = role(TextRole.SUBHEADING), color = FarmsyColors.ink)
            Text(text, style = role(TextRole.CAPTION), color = FarmsyColors.inkMuted)
        }
        action?.let { (label, run) -> PillButton(label, PillVariant.PRIMARY, PillSize.SMALL, onClick = run) }
    }
}

/// A followed farm: category tile, name, today's opening dot and distance.
@Composable
private fun FarmRow(pin: FarmPin, km: Double?, onClick: () -> Unit) {
    val open = FarmFilters.isOpenToday(pin.openingHours)
    Row(
        Modifier.fillMaxWidth().background(FarmsyColors.surface, CardShape).clickable(onClick = onClick).padding(Space.s3),
        verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(Space.s3),
    ) {
        Box(Modifier.size(48.dp).background(FarmsyColors.creamFill, RoundedCornerShape(Radius.thumb)), contentAlignment = Alignment.Center) {
            Text(pin.primaryCategory.emoji, fontSize = 22.sp)
        }
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            Text(pin.name, style = role(TextRole.SUBHEADING), color = FarmsyColors.ink, maxLines = 1, overflow = TextOverflow.Ellipsis)
            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                Box(Modifier.size(8.dp).background(if (open) FarmsyColors.vividPositive else FarmsyColors.hairline, CircleShape))
                Text(
                    stringResource(if (open) R.string.open_today else R.string.closed_today) + (km?.let { " · ${String.format("%.1f", it)} km" } ?: ""),
                    style = role(TextRole.CAPTION), color = FarmsyColors.inkMuted,
                )
            }
        }
        Icon(Icons.Filled.ChevronRight, null, tint = FarmsyColors.inkMuted, modifier = Modifier.size(18.dp))
    }
}

File: Farmsy/Features/Home/HomeScreen.swift
import SwiftUI
import CoreLocation

/// Home — "what should I know today?" Personal, not a directory: what is
/// available near you, the farms you follow, and what Plus would tell you.
/// Profile lives behind the person button in the header.
struct HomeScreen: View {
    @Environment(SessionStore.self) private var session
    @Environment(FarmsStore.self) private var farms
    @Environment(FavoritesStore.self) private var favorites
    @Environment(LocationManager.self) private var locationManager
    @Environment(\.shell) private var shell
    @Environment(\.requestAuth) private var requestAuth

    /// How far "near you" reaches. Shared with Shopping and the map's product tap.
    @AppStorage("searchRadiusKm") private var radiusKm = 15.0
    static let radiusChoices: [Double] = [5, 10, 15, 25, 50]

    @State private var catalogue = ShoppingItems.shared
    @State private var seasons = Seasons.shared
    @State private var recent = RecentReports.shared
    @State private var query = ""
    @State private var nearby: [ProductNearby] = []
    @State private var nearbyReady = false

    private var location: CLLocation? { locationManager.location }

    private var greeting: String {
        let hour = Calendar.current.component(.hour, from: Date())
        let name = session.profile?.firstName?.trimmingCharacters(in: .whitespaces) ?? ""
        let base = switch hour {
        case 5..<12: String(localized: "Good morning")
        case 12..<18: String(localized: "Good afternoon")
        default: String(localized: "Good evening")
        }
        return name.isEmpty ? base : "\(base), \(name)"
    }

    private var savedPins: [FarmPin] {
        farms.sortedByDistance(farms.pins.filter { favorites.isSaved($0.osmId) }, from: location)
    }

    var body: some View {
        ScrollView(showsIndicators: false) {
            VStack(alignment: .leading, spacing: 0) {
                band
                VStack(alignment: .leading, spacing: 0) {
                    greetingCard.padding(.top, -Space.s2)
                    if !session.hasFullAccess { plusRow }
                    availableSection
                    routeCard
                    thisWeekSection
                    yourFarmsSection
                    forYouSection
                }
                .padding(.horizontal, Space.s4)
                .padding(.bottom, TabBarInset.content)
            }
        }
        .background(Color.cream.ignoresSafeArea())
        .ignoresSafeArea(edges: .top)
        .task { await catalogue.loadIfNeeded() }
        .task { await seasons.loadIfNeeded() }
        .task { await farms.loadFlagsIfNeeded() }
        .task(id: session.hasFullAccess) { if session.hasFullAccess { await recent.refresh() } }
        .task(id: nearbyKey) { await computeNearby() }
        .onAppear { if location == nil { locationManager.request() } }
    }

    // MARK: - Header band

    private var band: some View {
        ZStack(alignment: .top) {
            AtmosphereBand()
            HStack {
                IconButton("person", label: String(localized: "Profile"), overMedia: true) {
                    shell.openProfile()
                }
                Spacer()
                Wordmark(onAtmosphere: true)
                Spacer()
                Menu {
                    ForEach(Self.radiusChoices, id: \.self) { km in
                        Button {
                            radiusKm = km
                        } label: {
                            if km == radiusKm {
                                Label("\(Int(km)) km", systemImage: "checkmark")
                            } else {
                                Text("\(Int(km)) km")
                            }
                        }
                    }
                } label: {
                    Image(systemName: "location")
                        .font(.system(size: 20, weight: .medium))
                        .foregroundStyle(Color.ink)
                        .frame(width: 44, height: 44)
                        .background(Color.white.opacity(0.92), in: Circle())
                        .shadow(color: Color.ink.opacity(0.10), radius: 12, y: 8)
                }
                .accessibilityLabel(String(localized: "Search radius"))
            }
            .padding(.horizontal, Space.s4)
            .padding(.top, 56)
        }
        .frame(height: 176)
    }

    // MARK: - Greeting

    private var greetingCard: some View {
        VStack(alignment: .leading, spacing: Space.s4) {
            VStack(alignment: .leading, spacing: Space.s1) {
                Text(greeting).role(.heading)
                Text(location == nil
                     ? String(localized: "Local food near you")
                     : String(localized: "Local food within \(Int(radiusKm)) km"))
                    .role(.bodySm, .inkMuted)
            }
            SearchField(text: $query, placeholder: String(localized: "Search farms or products")) {
                let q = query.trimmingCharacters(in: .whitespaces)
                guard !q.isEmpty else { return }
                farms.searchText = q
                shell.showTab(.map)
            }
            Button {
                Haptics.tap()
                farms.clearAllFilters()
                farms.filterOpenNow = true
                shell.showTab(.map)
            } label: {
                Label(String(localized: "Open now near me"), systemImage: "clock")
            }
            .buttonStyle(PillButtonStyle(.primary, size: .medium, block: true))
        }
        .card()
    }

    // MARK: - Plus

    /// The one obvious way in for a free member, right under the greeting. The
    /// lock card at the bottom sells a feature; this sells the membership.
    private var plusRow: some View {
        Button { Haptics.tap(); shell.openPlus(.homeRow) } label: {
            HStack(spacing: Space.s3) {
                Image(systemName: "sparkles")
                    .font(.system(size: 20, weight: .semibold))
                    .foregroundStyle(Color.farmGreen)
                VStack(alignment: .leading, spacing: 2) {
                    Text("Upgrade to Farmsy Plus").role(.subheading)
                    Text("Farmsy finds it, plans the route and tells you when it is fresh.").role(.caption, .inkMuted)
                }
                Spacer()
                Badge(text: "PLUS", fill: .vivid, ink: .ink)
            }
        }
        .buttonStyle(.plain)
        .card()
        .padding(.top, Space.s3)
    }

    // MARK: - Available near you

    private var nearbyKey: String {
        "\(location?.coordinate.latitude ?? 0),\(location?.coordinate.longitude ?? 0),\(radiusKm),\(catalogue.items.count),\(farms.flagsLoaded),\(farms.pins.count)"
    }

    private func computeNearby() async {
        guard let loc = location, farms.flagsLoaded, !catalogue.items.isEmpty, !farms.pins.isEmpty else { return }
        let items = catalogue.items, pins = farms.pins, produce = farms.produceByOsm
        let origin = loc.coordinate, radius = radiusKm
        nearby = await Task.detached(priority: .userInitiated) {
            FarmsStore.productsNearby(items: items, pins: pins, produce: produce, origin: origin, radiusKm: radius)
        }.value
        nearbyReady = true
    }

    @ViewBuilder
    private var availableSection: some View {
        SectionHeader(title: String(localized: "Available near you"),
                      action: (String(localized: "Shopping"), { shell.showTab(.shopping) }))
        if location == nil {
            locationCard
        } else if !nearbyReady {
            HStack(spacing: Space.s3) {
                ForEach(0..<3, id: \.self) { _ in
                    SkeletonBox(cornerRadius: Radius.tile).frame(width: 172, height: 120)
                }
            }
        } else if nearby.isEmpty {
            Text("No farm within \(Int(radiusKm)) km lists what it sells yet. Widen the radius from the header.")
                .role(.bodySm, .inkMuted)
                .card()
        } else {
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: Space.s3) {
                    ForEach(nearby.prefix(10)) { p in
                        productTile(p)
                    }
                }
                .padding(.trailing, Space.s4)
            }
            .padding(.trailing, -Space.s4)
        }
    }

    /// Plus: the newest "open" report within the radius that names this
    /// product, as "confirmed 18 min ago". Free tiles say nothing about timing.
    private func confirmedLine(_ itemId: String) -> String? {
        guard session.hasFullAccess, let location else { return nil }
        let hit = recent.near(location, radiusKm: radiusKm, pins: farms.pinsById)
            .filter { $0.report.status == .open && $0.report.products.contains(itemId) }
            .max { $0.report.createdAt < $1.report.createdAt }
        guard let hit else { return nil }
        let mins = max(0, Int(Date().timeIntervalSince(hit.report.createdAt) / 60))
        if mins < 60 { return String(localized: "Confirmed \(mins) min ago") }
        if mins < 60 * 36 { return String(localized: "Confirmed \(mins / 60) h ago") }
        return String(localized: "Confirmed \(mins / 1440) d ago")
    }

    private func productTile(_ p: ProductNearby) -> some View {
        Button {
            Haptics.tap()
            shell.openProduct(p.item.id)
        } label: {
            VStack(alignment: .leading, spacing: Space.s2) {
                ProductImage(slug: p.item.imageSlug, fallback: p.item.emoji, size: 64)
                Text(p.item.label).role(.subheading).lineLimit(1)
                Text("\(p.count) farms · \(p.nearestKm.formatted(.number.precision(.fractionLength(1)))) km")
                    .role(.caption, .inkMuted)
                    .lineLimit(1)
                if let line = confirmedLine(p.item.id) {
                    HStack(spacing: 4) {
                        Circle().fill(Color.vividPositive).frame(width: 6, height: 6)
                        Text(line).role(.caption, .positive).lineLimit(1)
                    }
                }
            }
            .frame(width: 156, alignment: .leading)
            .padding(Space.s4)
            .background(Color.surface, in: RoundedRectangle(cornerRadius: Radius.tile, style: .continuous))
        }
        .buttonStyle(.plain)
    }

    private var locationCard: some View {
        HStack(spacing: Space.s3) {
            Image(systemName: "location.slash")
                .font(.system(size: 20, weight: .medium))
                .foregroundStyle(Color.farmGreen)
            VStack(alignment: .leading, spacing: 2) {
                Text("Where are you?").role(.subheading)
                Text("Allow location to see what is for sale around you.").role(.caption, .inkMuted)
            }
            Spacer()
            Button(String(localized: "Allow")) { Haptics.tap(); locationManager.request() }
                .buttonStyle(PillButtonStyle(.primary, size: .small))
        }
        .card()
    }

    // MARK: - Route planner entry

    /// A second way into the route planner — the only other entrance is Shopping →
    /// "Build my route", easy to miss. Sits right under what's for sale nearby.
    private var routeCard: some View {
        Button { Haptics.tap(); shell.openTrips() } label: {
            HStack(spacing: Space.s4) {
                Image(systemName: "car.fill").font(.system(size: 20, weight: .semibold)).foregroundStyle(Color.farmGreen)
                VStack(alignment: .leading, spacing: 2) {
                    Text("Plan a farm route").role(.subheading)
                    Text("Pick your stops, Farmsy orders them and draws the road.").role(.caption, .inkMuted)
                }
                Spacer()
                if !session.hasFullAccess { Badge(text: "PLUS", fill: .vivid, ink: .ink) }
            }
        }
        .buttonStyle(.plain)
        .card()
    }

    // MARK: - This week

    /// Season news. Renders nothing until the calendar has loaded, so a missing
    /// endpoint costs a section, not a screen.
    @ViewBuilder
    private var thisWeekSection: some View {
        let picks = Array(seasons.thisMonth.prefix(6))
        if !picks.isEmpty {
            SectionHeader(title: String(localized: "In season now"),
                          action: (String(localized: "See all"), { shell.showTab(.discover) }))
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: Space.s3) {
                    ForEach(picks) { item in
                        seasonTile(item)
                    }
                }
                .padding(.trailing, Space.s4)
            }
            .padding(.trailing, -Space.s4)
        }
    }

    private func seasonTile(_ item: SeasonalItem) -> some View {
        Button {
            Haptics.tap()
            shell.openProduct(item.slug)
        } label: {
            VStack(alignment: .leading, spacing: Space.s2) {
                ZStack(alignment: .topTrailing) {
                    ProductImage(slug: item.imageSlug, fallback: item.emoji, size: 124, corner: Radius.tile)
                    if item.isPeak(month: seasons.month) {
                        Badge(text: String(localized: "PEAK"), fill: .vivid, ink: .ink).padding(Space.s2)
                    }
                }
                Text(item.label).role(.subheading).lineLimit(1)
                Text(String(localized: "Find it nearby")).role(.caption, .inkMuted)
            }
            .frame(width: 124, alignment: .leading)
            .padding(Space.s3)
            .background(Color.surface, in: RoundedRectangle(cornerRadius: Radius.card, style: .continuous))
        }
        .buttonStyle(.plain)
    }

    // MARK: - Your farms

    @ViewBuilder
    private var yourFarmsSection: some View {
        SectionHeader(title: String(localized: "Your farms"),
                      action: session.isAuthenticated && savedPins.count > 3
                          ? (String(localized: "See all"), { shell.showTab(.map) }) : nil)
        if !session.isAuthenticated {
            HStack(spacing: Space.s3) {
                Image(systemName: "heart")
                    .font(.system(size: 20, weight: .medium))
                    .foregroundStyle(Color.farmGreen)
                VStack(alignment: .leading, spacing: 2) {
                    Text("Keep your favourites").role(.subheading)
                    Text("Sign in to follow farms and see them here.").role(.caption, .inkMuted)
                }
                Spacer()
                Button(String(localized: "Sign in")) { Haptics.tap(); requestAuth() }
                    .buttonStyle(PillButtonStyle(.primary, size: .small))
            }
            .card()
        } else if savedPins.isEmpty {
            HStack(spacing: Space.s3) {
                Image(systemName: "heart")
                    .font(.system(size: 20, weight: .medium))
                    .foregroundStyle(Color.farmGreen)
                VStack(alignment: .leading, spacing: 2) {
                    Text("No farms followed yet").role(.subheading)
                    Text("Tap the heart on a farm to see it here, with today's opening.").role(.caption, .inkMuted)
                }
                Spacer()
            }
            .card()
        } else {
            VStack(spacing: Space.s2) {
                ForEach(savedPins.prefix(3)) { pin in
                    farmRow(pin)
                }
            }
        }
    }

    private func farmRow(_ pin: FarmPin) -> some View {
        let open = FarmFilters.isOpenToday(pin.openingHours)
        let km = pin.distance(from: location).map { $0 / 1000 }
        return Button {
            Haptics.tap()
            shell.openFarm(pin)
        } label: {
            HStack(spacing: Space.s3) {
                Text(pin.primaryCategory.emoji)
                    .font(.system(size: 22))
                    .frame(width: 48, height: 48)
                    .background(Color.creamFill, in: RoundedRectangle(cornerRadius: Radius.thumb, style: .continuous))
                VStack(alignment: .leading, spacing: 2) {
                    Text(pin.name).role(.subheading).lineLimit(1)
                    HStack(spacing: 6) {
                        Circle().fill(open ? Color.vividPositive : Color.hairline).frame(width: 8, height: 8)
                        Text(open ? String(localized: "Open today") : String(localized: "Closed today"))
                        if let km { Text(verbatim: "· \(km.formatted(.number.precision(.fractionLength(1)))) km") }
                    }
                    .role(.caption, .inkMuted)
                }
                Spacer(minLength: 0)
                Image(systemName: "chevron.right")
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(Color.inkMuted)
            }
            .padding(Space.s3)
            .background(Color.surface, in: RoundedRectangle(cornerRadius: Radius.card, style: .continuous))
        }
        .buttonStyle(.plain)
    }

    // MARK: - For you

    @ViewBuilder
    private var forYouSection: some View {
        SectionHeader(title: String(localized: "For you"))
        if session.hasFullAccess {
            Row(icon: "bell", title: String(localized: "Product alerts"),
                subtitle: String(localized: "Follow a product and hear when it turns up nearby")) {
                if let url = URL(string: "https://www.farmsy.app/alerts") { UIApplication.shared.open(url) }
            }
            .background(Color.surface, in: RoundedRectangle(cornerRadius: Radius.card, style: .continuous))
        } else {
            PlusLockCard(title: String(localized: "Alerts and live availability"),
                         text: String(localized: "Hear when strawberries turn up within your radius, and see how recently a farm was confirmed open."),
                         onUnlock: { shell.openPlus(.homeCard) })
        }
    }
}
